packages/data/src/components/with-select/index.js

```diff
@@ -9,6 +9,44 @@ import { createHigherOrderComponent } from '@wordpress/compose';
  * Internal dependencies
  */
 import { RegistryConsumer } from '../registry-provider';
+import { AsyncModeConsumer } from '../async-mode-provider';
+
+const waitingList = [];
+const componentsMap = new WeakMap();
+let isRunning = false;
+const runWaitingList = () => {
+	if ( waitingList.length === 0 ) {
+		isRunning = false;
+		return;
+	}
+	const nextComponent = waitingList.shift();
+	componentsMap.get( nextComponent )();
+	componentsMap.delete( nextComponent );
+	window.requestAnimationFrame( runWaitingList );
```

Comment:
I'm not sure a rAF is what you want here; this doesn't look like it should be tied to the screen's refresh rate.

What's the goal behind this wait? If you're just looking to defer things until the next run of the event loop, I'd use a simple lodash `defer` or the equivalent `setTimeout()` with a timeout of zero.

Do you instead want to make sure it doesn't all happen in one go and only when the browser is otherwise free to process things? If so, I'd use `requestIdleCallback` instead, which is only executed when the browser has spare time, with a polyfill or fallback of some sort for browsers that don't support it.

Reply:
That sounds like a good use-case for `requestIdleCallback`, then. Which as far as I can tell is what React's async mode uses as well, but I may have been looking at old code.

The motivation against using `requestAnimationFrame` is that you'd be doing non-animation work when the browser is getting ready to create a frame. This could lead to one of two things:
- If the work involved in each of these is complex, you could blow the 13ms budget and cause jank.
- If the work involved in each of these is not complex, you could be wasting time by taking `numberOfItemsInQueue * 13ms` to process the whole queue, when you could potentially be processing it all in much less time.

`requestIdleCallback` instead schedules things to be done when the browser expects it will have at least 50ms (I believe), which gives you a bigger budget for complex changes and allows for running multiple chunks of work in succession without waiting for another cycle if they're small enough. This could help in keeping changes coherent, too, instead of potentially having one component update per frame.

Reply:
Looks like `requestIdleCallback` is not supported by the browsers we need, and I'm not certain it has polyfills.

React uses `requestAnimationFrame` to provide some kind of polyfill.

https://github.com/facebook/react/blob/653bc582f94e955b175d42836479ee8b9eeeeeaf/packages/scheduler/src/Scheduler.js

Reply:
I'd be OK using rIC where supported and just having a polyfill as simple as `rIC = ( ...args ) => requestAnimationFrame( ...args )` elsewhere, but I'll leave this up to you. As long as the numbers improve, that's what matters :)

Reply:
Does this mean between each invocation, it will wait 50ms at least? If it's the case I think `requestAnimationFrame` could be better? and each component's withSelect is fast, it's the accumulation that is not.

Reply:
I replaced with requestIdleCallback but still not certain here :)

Reply:
Actually, that test I ran doesn't match what you were doing, which is scheduling an idle callback inside an existing one. Give me one second for an updated test :)

Reply:
Right, if scheduled inside an idle callback, it'll wait until the next one:

```js
let count = 50;
const callback = () => {
  console.log(performance.now());
  count--;
  if(count > 0) {
    requestIdleCallback(callback);
  }
}
```

```
2769447.6099999156
2769500.139999902
2769552.7900001034
2769607.8449999914
2769657.995000016
2769710.170000093
2769764.2500000075
2769816.4150000084
2769870.9849999286
2769925.0350000802
2769976.1999999173
2770026.4850000385
2770080.4949998856
2770133.785000071
2770188.8800000306
2770241.1799998954
2770295.9100001026
2770350.0600000843
2770403.644999955
2770458.765000105
2770508.930000011
2770559.2199999373
2770611.3899999764
2770663.679999998
2770718.355000019
2770769.7350000963
2770821.1699998938
2770875.3200001083
2770926.1449999176
2770978.550000116
2771031.154999975
2771081.3299999572
2771133.7649999186
2771188.8749999925
2771241.65500002
2771295.230000047
2771347.2299999557
2771397.4450000096
2771449.479999952
2771503.4099998884
2771558.505000081
2771610.5049999896
2771664.439999964
2771714.894999983
2771765.284999972
2771820.390000008
2771875.449999934
2771925.7950000465
2771975.9750000667
2772028.2199999783
```

The correct way is indeed to use `timeRemaining`:

```js
let count = 50;
const callback = (deadline) => {
  while (deadline.timeRemaining() > 0 && count > 0) {
    console.log(performance.now());
    count--;
  }
  if(count > 0) {
    requestIdleCallback(callback);
  }
}
```

```
13962.54500001669
13962.680000113323
13962.765000062063
13962.855000048876
13962.945000035688
13963.025000179186
13963.100000051782
13963.190000038594
13963.270000182092
13963.355000130832
13963.4350000415
13963.515000184998
13963.590000057593
13963.715000078082
13963.81500014104
13963.905000127852
13963.98500003852
13964.060000143945
13964.145000092685
13964.225000003353
13964.305000146851
13964.385000057518
13964.460000162944
13964.545000111684
13964.62000021711
13964.695000089705
13964.77000019513
13964.845000067726
13964.930000016466
13965.005000121891
13965.120000066236
13965.220000129193
13965.315000154078
13965.395000064746
13965.51000000909
13965.620000148192
13965.70000005886
13965.780000202358
13965.85000003688
13965.940000023693
13966.04000008665
13966.130000073463
13966.21000021696
13966.300000203773
13966.394999995828
13966.475000139326
13966.555000049993
13966.63000015542
13966.71500010416
13966.775000095367
```

Reply:
I'm feeling for a start, it's probably safer to use `requestAnimationFrame` otherwise we'll have to build two different behaviors depending on which API is available.

Reply:
Applying that to your code, I'd probably do something like:

```js
const runWaitingList = ( deadline ) => {
  if ( waitingList.length === 0 ) {
    isRunning = false;
    return;
  }

  do {
    const nextElement = waitingList.shift();
    elementsMap.get( nextElement )();
    elementsMap.delete( nextElement );
  } while ( waitingList.length > 0 && deadline && deadline.timeRemaining &&
      deadline.timeRemaining() > 0 )

  if ( waitingList.length > 0 ) {
    requestFrame( runWaitingList );
  }
};
```

Unless I'm missing something, that should handle both the rIC and rAF cases well.

Reply:
> I'm feeling for a start, it's probably safer to use `requestAnimationFrame` otherwise we'll have to build two different behaviors depending on which API is available.

Fair enough, I can see the logic in that. Let me know once you've made the changes you opted for, in that case, and I'll get right to those performance measurements.

Reply:
While your code is good, I still have more confidence personally in using requestAnimationFrame for this reason:

It seems that in your code the queue will be cleared by batch which mean it will have two consequences:

 - While executing a single batch (I guess it's 50ms duration), we can't perform incoming higher-priority work (typing)
 - After each batch, we'll wait another 50ms for the next batch.

Anyway I updated using rAF.

Reply:
> While your code is good, I still have more confidence personally in using requestAnimationFrame for this reason:
> 
> It seems that in your code the queue will be cleared by batch which mean it will have two consequences:
> 
> * While executing a single batch (I guess it's 50ms duration), we can't perform incoming higher-priority work (typing)
> * After each batch, we'll wait another 50ms for the next batch.
> 
> Anyway I updated using rAF.

No problem, I'll test as-is (using rAF) :) Again, as the improvements are there and it works correctly, it's already a win!

I just want to clarify that higher-priority work isn't an issue in the rIC case; that's the point of checking `timeRemaining()` after each item, and moving onto a new batch if there's no time left. We're trusting the browser to tell us there's no time left if more important work needs to be done.

As for waiting 50ms, yes, that's true, the wait is longer that the 13ms of a rAF. However, timing is often about waiting for the *right time* to do something, even if it takes a little longer. That's exactly what rIC tries to do, by using timing knowledge that's internal to the browser. Besides, in the rAF case, you'll incur the 13ms penalty between every item, rather than 50ms between batches in rIC. And since you expect each component's `withSelect` to be fast (which I assume to mean significantly less than 12ms), that likely means it'll be slower in aggregate, since the average batch would only need to have around 4 items for the rIC case to be faster.

Reply:
Thanks for the clarifications :) I'm convinced now

Reply:
Updated :)

Reply:
Sweet, thank you! :D I should be able to get you some numbers today! :)

Reply:
React uses `requestAnimationFrame` to implement polyfill for `requestIdleCallback`, which would otherwise be exactly the function they want to call.

The goal is to run code as soon as possible after a frame is painted. That's achieved with:
1. Scheduling a callback with `requestAnimationFrame`. But that alone schedules the callback right **before** a frame is painted, the opposite of what we want to achieve. That's why this callback posts a message to the window.
2. Then there is a `window.onmessage` handler that does the real work. If a message is posted inside `requestAnimationFrame`, it's guaranteed to be processed ASAP, but not before the frame is painted. It'll be the first thing the browser will do after painting! Mission accomplished.

`requestAnimationFrame` callback, i.e., running code before frame paint, is useful for **writing** to DOM effectively, without causing layout trashing.

Running code after paint is useful for effective **reading** from DOM, without causing layout trashing.

React Hooks introduce new APIs for these purposes. [useEffect](https://reactjs.org/docs/hooks-reference.html#useeffect) is ran before paint and is good for DOM writing, and [useLayoutEffect](https://reactjs.org/docs/hooks-reference.html#uselayouteffect) is ran after paint and is good for DOM reading. Or is it the other way around? I'm not sure :smile:

Calypso has an [afterLayoutFlush](https://github.com/Automattic/wp-calypso/tree/master/client/lib/after-layout-flush) utility function that follows very closely the React implementation to schedule stuff after paint. But it uses `setTimeout(0)` instead of `postMessage`, because I didn't know the hairy details a year ago. `setTimeout(0)` has the disadvantage that it can skip many frames before running, e.g., when scrolling.